In [1]:
#import all libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Creating a folder path
path = r'C:\Users\Lenovo\Desktop\Career Foundry\28-06-2024 Instacart Basket Analysis - Saswata Biswas'

In [3]:
#Using the os.path.function and importing the order file
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)


In [4]:
#Using the os.path.function and importing the products file
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
#checking the orders file
df_ords.columns

Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order'],
      dtype='object')

In [6]:
#Checking the products file
df_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

In [7]:
#Removing the Eval Set from Orders File
df_ords.drop(columns=['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [8]:
#Making a new dataframe instead of overwriting
df_ords_new = df_ords.drop(columns=['eval_set'])

In [9]:
#To see missing variabales in a column
df_ords['days_since_prior_order'].value_counts (dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [10]:
#To rename a column in the newly created df, this is the command:
df_ords_new.rename(columns = {'order_dow' : 'orders_days_of_week'}, inplace = True)

In [11]:
df_ords_new.head()

order_id  user_id  order_number  orders_days_of_week  order_hour_of_day  \
0   2539329        1             1                    2                  8   
1   2398795        1             2                    3                  7   
2    473747        1             3                    3                 12   
3   2254736        1             4                    4                  7   
4    431534        1             5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [12]:
#Changing the data types:
df_ords_new['order_id'] = df_ords_new['order_id'].astype('str')

In [13]:
#checking the new type
df_ords_new['order_id'].dtype

dtype('O')

In [14]:
#importing data set departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [15]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [16]:
#Transposing the department file, changing from rows to columns
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [17]:
#Creating the trasponsed dataframe
df_dep_t = df_dep.T

In [18]:
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [19]:
#giving an index to a file that does not have one:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [20]:
new_header = df_dep_t.iloc[0]

In [21]:
new_header

0    department
Name: department_id, dtype: object

In [22]:
#creating a new dataframe that only copies over rows beyond the first row:
df_dep_t_new = df_dep_t[1:]

In [23]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [24]:
# to add new header:
df_dep_t_new.columns = new_header

In [25]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [26]:
#transforming the department csv into a dictionary:
data_dict = df_dep_t_new.to_dict('index')

In [27]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [28]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [30]:
#creating a subset frrom the snack department
df_snacks =  df_prods[df_prods['department_id']==19]

In [31]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

#The other Wrangling procedures in the exercise

Question 02. Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [34]:
#let us change the data type of order_number
df_ords_new['order_number'] = df_ords_new['order_number'].astype('str')

In [35]:
#checking the new type
df_ords_new['order_id'].dtype

dtype('O')

Question 3. Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe

In [37]:
#Order hour of day to be changed to order_time
#df_ords_new.rename(columns = {'order_hour_of_day' : 'orders_time'}, inplace = False)
#Please note that I did not run this command in case I am wrong

Question 4: Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [39]:
#We need to find the counts of the column order_hour_of_day
df_ords_new['order_hour_of_day'].value_counts (dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

In [40]:
# Busiest hour of the day is at 10 am

Question 05. Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary

In [42]:
print(data_dict.get('4'))

{'department': 'produce'}


Question 06. The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.

In [44]:
#The breakfast is department id 14, we will name the subset as df_breakfast
df_breakfast =  df_prods[df_prods['department_id']==14]

In [45]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

Question 07: They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.

In [47]:
#Let us name this subset as parties Department IDs of alcohol is 5, deli is 20, beverages is 7, meat/seafood is 12
df_parties = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

Question 08: It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

In [49]:
df_parties.shape

(7650, 5)

Answer: it has 7650 rows and 5 columns

Question 09: Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [114]:
df_ords_new.head()

order_id  user_id order_number  orders_days_of_week  order_hour_of_day  \
0  2539329        1            1                    2                  8   
1  2398795        1            2                    3                  7   
2   473747        1            3                    3                 12   
3  2254736        1            4                    4                  7   
4   431534        1            5                    4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [120]:
#let us make a subset of the user id 1:
df_user_id_1=df_ords_new.loc[df_ords_new['user_id']==1]

In [126]:
df_user_id_1.head(20)

order_id  user_id order_number  orders_days_of_week  order_hour_of_day  \
0   2539329        1            1                    2                  8   
1   2398795        1            2                    3                  7   
2    473747        1            3                    3                 12   
3   2254736        1            4                    4                  7   
4    431534        1            5                    4                 15   
5   3367565        1            6                    2                  7   
6    550135        1            7                    1                  9   
7   3108588        1            8                    1                 14   
8   2295261        1            9                    1                 16   
9   2550362        1           10                    4                  8   
10  1187899        1           11                    4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

This shows the user has only ordered 11 times. Thursday, friday, saturday there are no orders

12. Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder : Done below

In [133]:
df_ords_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

13. Export the df_dep_t_new dataframe as “departments_wrangled.csv” in your “Prepared Data” folder so that you have a “.csv” file of your departments data in the correct format. Done below

In [138]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))